hazelcast/src/main/java/com/hazelcast/internal/partition/operation/PartitionReplicaSyncRequest.java

```diff
@@ -81,42 +82,60 @@ public void beforeRun() throws Exception {
     }
 
     @Override
-    public void run() throws Exception {
-        NodeEngineImpl nodeEngine = (NodeEngineImpl) getNodeEngine();
-        InternalPartitionServiceImpl partitionService = (InternalPartitionServiceImpl) nodeEngine.getPartitionService();
-        int partitionId = getPartitionId();
-        int replicaIndex = getReplicaIndex();
-
+    public void run() {
+        ILogger logger = getLogger();
+        InternalPartitionServiceImpl partitionService = getService();
         if (!partitionService.areMigrationTasksAllowed()) {
-            ILogger logger = getLogger();
             if (logger.isFinestEnabled()) {
-                logger.finest("Migration is paused! Cannot run replica sync -> " + toString());
+                logger.finest("Migration is paused! Cannot process request. partitionId="
+                        + getPartitionId() + ", replicaIndex=" + getReplicaIndex() + ", namespaces=" + namespaces);
             }
             sendRetryResponse();
             return;
         }
 
-        if (!preCheckReplicaSync(nodeEngine, partitionId, replicaIndex)) {
+        final int permits = acquirePermits();
+        if (permits == 0) {
+            if (logger.isFinestEnabled()) {
+                logger.finest("Not enough permits available! Cannot process request. partitionId="
+                        + getPartitionId() + ", replicaIndex=" + getReplicaIndex() + ", namespaces=" + namespaces);
+            }
+            sendRetryResponse();
             return;
         }
 
-        try {
-            PartitionReplicationEvent event = new PartitionReplicationEvent(partitionId, replicaIndex);
-            if (allNamespaces.remove(NonFragmentedServiceNamespace.INSTANCE)) {
-                Collection<Operation> operations = createNonFragmentedReplicationOperations(event);
-                sendOperations(operations, NonFragmentedServiceNamespace.INSTANCE);
+        sendOperationsForNamespaces(permits);
+        if (!namespaces.isEmpty()) {
+            if (logger.isFinestEnabled()) {
+                logger.finest("Not enough permits available! Cannot process request. partitionId="
+                        + getPartitionId() + ", replicaIndex=" + getReplicaIndex() + ", namespaces=" + namespaces);
             }
+            sendRetryResponse();
```

Comment:
Will this retry all of the namespaces (even if some operations have been sent)? Can the retry response be received by the caller before the partition data operations?

Reply:
`sendOperationsForNamespaces()` removes namespaces from the list after sending data. So `sendRetryResponse ()` will only retry the remaining namespaces.